examples/01_basics/e5_reatil_cart.py

```diff
@@ -0,0 +1,116 @@
+"""
+A retail cart example that:
+- Tracks the real-time total of a cart (as items are added and removed)
+- Gives a 10% discount if the purchase is made within XX minutes
+- Tracks the total sales and number of purchases
+"""
+
+from datetime import datetime, timedelta
+
+import csp
+import csp.showgraph
+from csp import ts
+
+
+PRODUCTS = {
+    "X": 20.0,
+    "Y": 50.0,
+}
+
+
+class Cart(csp.Struct):
+    product: str
+    qty: int
+    add: bool
+    purchase: bool
+
+
+@csp.node
+def track_discount(event: ts[Cart]) -> ts[float]:
+    with csp.alarms():
+        alarm = csp.alarm(bool)
+
+    with csp.state():
+        discount = 0.9
+
+    with csp.start():
+        csp.schedule_alarm(alarm, timedelta(seconds=5), True)
+
+    if csp.ticked(alarm):
+        discount = 1
+
+    return discount
+
+
+@csp.node
+def update_cart(event: ts[Cart], discount: ts[float]) -> csp.Outputs(total = ts[float], items = ts[int]):
+    with csp.state():
+        cart_total = 0.0
+        cart_items = 0
+
+    if csp.ticked(event):
+        if event.add:
+            cart_total += PRODUCTS[event.product] * event.qty
+            cart_items += event.qty
+        else:
+            cart_total -= PRODUCTS[event.product] * event.qty
+            cart_items += event.qty
+
+    final_total = cart_total * discount
+
+    if event.purchase:
+        cart_total = 0.0
+        cart_items = 0
+
+    csp.output(total=final_total, items=cart_items)
+
+
+@csp.node
+def track_purchases(event: ts[Cart], cart_total: ts[float]) -> csp.Outputs(sale = ts[float], qty = ts[int]):
+    with csp.state():
+        purchases = 0
+        total_sales = 0.0
+
+    purchases += 1
```

Comment:
should this tick any time event ticks or only when cart_total ticks?
in general I try to be pretty liberal with adding csp.ticked() checks in nodes

Reply:
not sure why this one was marked resolved

Reply:
It needs to tick with all events where `purchase=True`.

I marked it as resolved because I added a `csp.ticked()` [check here](https://github.com/Point72/csp/pull/301/files#diff-91402a82bda892ae959c92877608c1884e0dc850c324f0235414077bfb4bcd06R85-R87).

I think we lost the exact line references because the file was re-named to fix the typo.